app/src/titlebarwidget.cpp

```diff
@@ -0,0 +1,210 @@
+#include <QVBoxLayout>
+#include <QLabel>
+#include <QIcon>
+#include <QToolButton>
+#include <QResizeEvent>
+#include <QStyle>
+#include <QPainter>
+#include <QMenu>
+#include <QDebug>
+
+#include "platformhandler.h"
+#include "buttonappearancewatcher.h"
+
+TitleBarWidget::TitleBarWidget(QWidget* parent)
+    : QWidget(parent)
+{
+
+    QVBoxLayout* vLayout = new QVBoxLayout();
+
+    vLayout->setContentsMargins(3,4,3,4);
+    vLayout->setSpacing(0);
+
+    vLayout->addWidget(createCustomTitleBarWidget(this));
+
+    setLayout(vLayout);
+}
+
+TitleBarWidget::~TitleBarWidget()
+{
+}
+
+QWidget* TitleBarWidget::createCustomTitleBarWidget(QWidget* parent)
+{
+    bool isDarkmode = PlatformHandler::isDarkMode();
+    QWidget* containerWidget = new QWidget(parent);
+
+    QHBoxLayout* containerLayout = new QHBoxLayout(parent);
+
+    mCloseButton = new QToolButton(parent);
+
+    mCloseButton->setStyleSheet(flatButtonStylesheet());
+
+    QSize iconSize = QSize(14,14);
+    QSize padding = QSize(2,2);
+
+    IconResource closeButtonRes;
+    closeButtonRes.lightMode = QIcon("://icons/themes/playful/window/window-close-button-normal.svg");
+    closeButtonRes.darkMode = QIcon("://icons/themes/playful/window/window-close-button-normal-darkm.svg");
+
+    QIcon closeIcon = closeButtonRes.iconForMode(isDarkmode);
+
+    IconResource closeHoverButtonRes;
+    closeHoverButtonRes.lightMode = QIcon("://icons/themes/playful/window/window-close-button-active.svg");
+    closeHoverButtonRes.darkMode = closeHoverButtonRes.lightMode;
+
+    mCloseButton->setIcon(closeIcon);
+    mCloseButton->setIconSize(iconSize);
+    mCloseButton->setFixedSize(iconSize + padding);
+    mCloseButton->installEventFilter(new ButtonAppearanceWatcher(closeButtonRes,
+                                                                 closeHoverButtonRes,
+                                                            this));
+
+    connect(mCloseButton, &QToolButton::clicked, this, &TitleBarWidget::closeButtonPressed);
+
+    IconResource dockButtonRes;
+    dockButtonRes.lightMode = QIcon("://icons/themes/playful/window/window-float-button-normal.svg");
+    dockButtonRes.darkMode = QIcon("://icons/themes/playful/window/window-float-button-normal-darkm.svg");
+
+    IconResource dockHoverButtonRes;
+    dockHoverButtonRes.lightMode = QIcon("://icons/themes/playful/window/window-float-button-active.svg");
+    dockHoverButtonRes.darkMode = dockHoverButtonRes.lightMode;
+
+    mDockButton = new QToolButton(parent);
+
+    QIcon dockIcon = dockButtonRes.iconForMode(isDarkmode);
+    mDockButton->setIcon(dockIcon);
+    mDockButton->setStyleSheet(flatButtonStylesheet());
+
+    mDockButton->setIconSize(iconSize);
+    mDockButton->setFixedSize(iconSize + padding);
+    mDockButton->installEventFilter(new ButtonAppearanceWatcher(dockButtonRes, dockHoverButtonRes, this));
+
+    connect(mDockButton, &QToolButton::clicked, this, [this] {
+       emit undockButtonPressed();
```

Comment:
This can use the classic connection syntax, too, just like the closeButtonPressed connection above.

Reply:
Fixed